src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,900 @@
+
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "Unable to send sendMissingTrustedTimeSource event [endpointId=%d]", endpointId);
+        return false;
+    }
+
+    // TODO re-schedule event for after min 1hr if TTS is null or cannot be reached
+    ChipLogProgress(Zcl, "Emit sendMissingTrustedTimeSource event [endpointId=%d]", endpointId);
+    return true;
+}
+
+TimeSynchronizationServer TimeSynchronizationServer::mTimeSyncInstance;
+
+TimeSynchronizationServer & TimeSynchronizationServer::Instance()
+{
+    return mTimeSyncInstance;
+}
+
+void TimeSynchronizationServer::Init()
+{
+    mTimeSyncDataProvider.Init(Server::GetInstance().GetPersistentStorage());
+
+    TimeSynchronization::Structs::TrustedTimeSourceStruct::Type tts;
+    (CHIP_NO_ERROR == mTimeSyncDataProvider.LoadTrustedTimeSource(tts)) ? (void) mTrustedTimeSource.SetNonNull(tts)
+                                                                        : mTrustedTimeSource.SetNull();
+    for (size_t i = 0; i < mTimeZoneList.size(); i++)
+    {
+        memset(mNames[i].name, 0, sizeof(mNames[i].name));
+        mTimeZoneList[i] = { 0, 0, Optional<chip::CharSpan>(chip::CharSpan(mNames[i].name, sizeof(mNames[i].name))) };
+    }
+    mTimeSyncDataProvider.LoadTimeZone(mTimeZoneList, mTimeZoneListSize);
+    mTimeZoneListSize = (mTimeZoneListSize == 0) ? 1 : mTimeZoneListSize; // initialize default value
+    mTimeSyncDataProvider.LoadDSTOffset(mDstOffsetList, mDstOffsetListSize);
+    // TODO if trusted time source is available schedule a time read
+    if (!mTrustedTimeSource.IsNull())
+    {
+    }
+    System::Clock::Microseconds64 utcTime;
+    if (System::SystemClock().GetClock_RealTime(utcTime) == CHIP_NO_ERROR)
+    {
+        mGranularity = TimeSynchronization::GranularityEnum::kMinutesGranularity;
+    }
+    else
+    {
+        mGranularity = TimeSynchronization::GranularityEnum::kNoTimeGranularity;
+    }
+}
+
+CHIP_ERROR TimeSynchronizationServer::SetTrustedTimeSource(
+    DataModel::Nullable<TimeSynchronization::Structs::TrustedTimeSourceStruct::Type> tts)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    mTrustedTimeSource = tts;
+    if (!mTrustedTimeSource.IsNull())
+    {
+        err = mTimeSyncDataProvider.StoreTrustedTimeSource(mTrustedTimeSource.Value());
+    }
+    else
+    {
+        err = mTimeSyncDataProvider.ClearTrustedTimeSource();
+    }
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationServer::SetDefaultNTP(DataModel::Nullable<chip::CharSpan> & dntp)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    if (!dntp.IsNull())
+    {
+        err = mTimeSyncDataProvider.StoreDefaultNtp(dntp.Value());
+    }
+    else
+    {
+        err = mTimeSyncDataProvider.ClearDefaultNtp();
+    }
+    return err;
+}
+
+CHIP_ERROR TimeSynchronizationServer::SetTimeZone(DataModel::DecodableList<TimeSynchronization::Structs::TimeZoneStruct::Type> tzL)
+{
+    size_t items;
+    ReturnErrorOnFailure(tzL.ComputeSize(&items));
```

Comment:
Does it matter that it returns INVALID_COMMAND? We are only interested in ResourceExhausted and ConstraintError here according to spec. I don't think changing to Status is better because all checks inside this function return CHIP_ERROR and I don't see any benefit in converting those.

Reply:
> Does it matter that it returns INVALID_COMMAND? 

Well, the spec requires INVALID_COMMAND in this case, so this particular implementation is not spec-compliant.